### PR TITLE
Expression evaluator

### PR DESCRIPTION
**Pull request by tlemo**
A basic expression evaluator which can be used on top of an fusion IR:

```
  Fusion fusion;
  FusionGuard fg(&fusion);
  ... create IR ...

  // 1. Create an evaluation context
  EvaluationContext eval_context(&fusion);

  // 2. Bind values
  eval_context.bind(tv0->getRootDomain()->axis(0)->extent(), 100);

  // 3. Evaluate values
  const auto result = ExpressionEvaluator::evaluate(val, eval_context);
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 1]
```diff
@@ -95,6 +95,29 @@ struct TORCH_CUDA_API Statement {
...
 
+  // Replacement for static_cast<T*>(ptr): ptr->as<T>()
+  template <class T>
+  T* as() {
+#ifdef NDEBUG
+    auto downcast_ptr = static_cast<T*>(this);
+#else
+    auto downcast_ptr = dynamic_cast<T*>(this);
+#endif
+    assert(downcast_ptr != nullptr);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: nitpick, not sure if we intended to assert for `NDEBUG`.
If so we should change this to `TORCH_INTERNAL_ASSERT`, otherwise, we can move this into the #else block above to avoid confusion.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 2]
```diff
@@ -95,6 +95,29 @@ struct TORCH_CUDA_API Statement {
...
+  }
+
+  template <class T>
+  const T* as() const {
+#ifdef NDEBUG
+    auto downcast_ptr = static_cast<const T*>(this);
+#else
+    auto downcast_ptr = dynamic_cast<const T*>(this);
+#endif
+    assert(downcast_ptr != nullptr);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: same here.

File: torch/csrc/jit/codegen/cuda/expr_evaluator.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+void ExpressionEvaluator::handle(const NamedScalar* i) {
+  // nothing to do, leave the result "unknown"
+}
+
+void ExpressionEvaluator::handle(const UnaryOp* uop) {
+  const auto in = evaluate(uop->in(), context_);
+  if (in.has_value()) {
+    switch (uop->getUnaryOpType()) {
+      case UnaryOpType::Neg:
+        result_ = Int::ScalarType(!*in);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/expr_evaluator.cpp]: `Neg` mapping to `!` looks dangerous.
Can we have add a test with `Neg` on integer as well as bool to confirm the behavior?

File: torch/csrc/jit/codegen/cuda/expr_evaluator.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        break;
+      case BinaryOpType::Div:
+        TORCH_CHECK(*rhs != 0);
+        result_ = *lhs / *rhs;
+        break;
+      case BinaryOpType::Mod:
+        TORCH_CHECK(*rhs != 0);
+        result_ = *lhs % *rhs;
+        break;
+      case BinaryOpType::LT:
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/expr_evaluator.cpp]: looks like `LT` and `And` is not getting tested.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
LGTM. Please address review comments.

File: test/cpp/jit/test_gpu.cpp [hunk 5]
```diff
@@ -49,6 +62,174 @@ void testGPU_FusionDispatch() {
...
+  tv2->axis(-1)->parallelize(ParallelType::TIDx);
+  tv3->axis(-1)->parallelize(ParallelType::TIDx);
+
+  // 1. Create an evaluation context
+  EvaluationContext eval_context(&fusion);
+
+  // 2. Bind values
+  //
+  // IMPORTANT:
+  // a. The bindings are only as stable as the Vals are in the fusion graph
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:147]: We could enforce that the evaluation is only valid after lowering the code. We could do this by running through all input TensorViews and making sure their domain extents are either constant or a named scalar.

File: test/cpp/jit/test_gpu.cpp [hunk 6]
```diff
@@ -49,6 +62,174 @@ void testGPU_FusionDispatch() {
...
+  checkIntValue(&eval_context, tv4->axis(1)->rawExtent(), 127);
+
+  TORCH_CHECK(tv5->domain()->nDims() == 1);
+  checkIntValue(&eval_context, tv5->axis(0)->rawExtent(), 16383);
+
+  TORCH_CHECK(tv6->domain()->nDims() == 3);
+  checkIntValue(&eval_context, tv6->axis(0)->rawExtent(), 26);
+  checkIntValue(&eval_context, tv6->axis(1)->rawExtent(), 5);
+  checkIntValue(&eval_context, tv6->axis(2)->rawExtent(), 127);
+}
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:232]: Really excited about the implications of being able to do this :-)

File: torch/csrc/jit/codegen/cuda/dispatch.h [hunk 7]
```diff
@@ -227,13 +227,13 @@ struct TORCH_CUDA_API OptInConstDispatch {
...
-    AT_ERROR("Handle not overriden for ForLoop.");
+    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for ForLoop.");
   }
   virtual void handle(const Allocate* const) {
-    AT_ERROR("Handle not overriden for Allocate.");
+    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for Allocate.");
   }
   virtual void handle(const IfThenElse* const) {
-    AT_ERROR("Handle not overriden for IfThenElse.");
+    TORCH_INTERNAL_ASSERT(false, "Handle not overriden for IfThenElse.");
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/dispatch.h:236]: Thanks for the clean up

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Can we do an explicit RTTI check to make sure the cast is valid?

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 8]
```diff
@@ -277,6 +277,9 @@ struct TORCH_CUDA_API IterDomain : public Val {
     return start_;
   }
   Val* extent() const;
+  Val* rawExtent() const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:280]: nit: noexcept?

**Review by csarofeen — CHANGES_REQUESTED** [review]
Only thing that concerns me is `as()` everything else looks good to me.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 1 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: @jjsjann123 : I moved the check inside the #else path and replaced it with TORCH_INTERNAL_ASSERT. There's a small problem the TORCH_INTERNAL_ASSERT (and TORCK_CHECK) - they report failures as exceptions. Which means that it's problematic to have anything which uses them in destructors.

> Can we do an explicit RTTI check to make sure the cast is valid?

Yes, that's the dynamic_cast<> path in debug builds (no check in release builds since this is a replacement for static_cast)

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 2 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Done.

File: torch/csrc/jit/codegen/cuda/expr_evaluator.cpp (hunk 3 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/expr_evaluator.cpp]: Hmm, I don't know what was I thinking (probably read Not instead of Neg). Fixed.

File: torch/csrc/jit/codegen/cuda/expr_evaluator.cpp (hunk 4 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/expr_evaluator.cpp]: Good catch. I meant to remove the support for LT since we only support integral results.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h (hunk 8 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:280]: As an optimization it's not helping much (most every C++ compiler would infer this from the definition, which is even inline here)

I'm a bit concerned about adding noexcept everywhere since it can actually add an overhead (compiler must generate code to check & enforce it unless it can prove it's not needed). 

I can add it if you insist, although it creates a slippery slope: in the absence of a detailed set of guidelines, the most it's used the more it encourages overuse, which in turn creates more problems than it solves (in particular, the unfortunate choice to have TORCH_CHECK report failures as exceptions makes it hard to predict where exceptions can originate from  a cursory review)

TL;DR: noexcept is a interface contract specification, not an optimization hint.

File: test/cpp/jit/test_gpu.cpp (hunk 5 above)
Comment (tlemo) [reply test/cpp/jit/test_gpu.cpp:147]: Why? What's wrong with evaluating expressions before lowering?

**Review by jjsjann123 — APPROVED** [review]
My requests are addressed.

**Review by csarofeen — APPROVED** [review]
